Add new exception when attempting to connect to MySQL running on localhost.

ruyrocha opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created to address #50824.

 when there is a MySQL server running on localhost, and the gem translates the hostname to a socket.

### Detail

This Pull Request changes the raised exception while attempting to connect to a MySQL server running on localhost, which is the case for a container running on local machine on default port.

Turns out the MySQL client library translates the hostname `localhost` to a socket, but there is the server up and running.

### Additional information

The new exception instructs you to use the IP address instead of the hostname when using the MySQL adapter.

FYI, it doesn't happen when using Trilogy as the adapter to connect to MySQL servers.

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [ ] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [ ] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

rafaelfranca commented:
> Turns out the MySQL client library translates the hostname `localhost` to a socket, but there is the server up and running.

Where is this happening? I could not find anything in mysql2 that does that.

ruyrocha commented:
@rafaelfranca to add more context, I'll paste the contents of [this comment](https://github.com/rails/rails/issues/50824#issuecomment-1915442937) here:

 The issue was firstly [reported to MySQL team on 2007](https://bugs.mysql.com/bug.php?id=31577), and they claim it's not a bug.

We can see from [their docs](https://dev.mysql.com/doc/refman/8.0/en/connecting.html):

>  Only connection options that are relevant to the selected transport protocol are used or checked. Other connection options are ignored. For example, with --host=localhost on Unix, the client attempts to connect to the local server using a Unix socket file, even if a --port or -P option is given to specify a TCP/IP port number.
>
> To ensure that the client makes a TCP/IP connection to the local server, use --host or -h to specify a host name value of 127.0.0.1 (instead of localhost), or the IP address or name of the local server. You can also specify the transport protocol explicitly, even for localhost, by using the --protocol=TCP option.

And the reproduction steps are straightforward:
```shell
ruy@devastator ~/G/r/activerecord (main)> mysql -h 127.0.0.1 -u root
ERROR 2002 (HY000): Can't connect to MySQL server on '127.0.0.1' (115)
ruy@devastator ~/G/r/activerecord (main) [1]> podman start mariadb
mariadb
ruy@devastator ~/G/r/activerecord (main)> mysql -h 127.0.0.1 -u root
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 3
Server version: 11.2.2-MariaDB-1:11.2.2+maria~ubu2204 mariadb.org binary distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> ^DBye
ruy@devastator ~/G/r/activerecord (main)> mysql -h localhost -u root
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)
ruy@devastator ~/G/r/activerecord (main) [1]>
```

As stated on the testing script for #50824, such behavior cannot be reprouced if Trilogy is used instead of MySQL adapter.

rafaelfranca commented on activerecord/lib/active_record/errors.rb:
```diff
@@ -88,6 +88,16 @@ def set_pool(connection_pool)
 
+  # Raised when a connection could not be obtained to the server running
+  # on localhost because it attempts to connect through a socket.
+  class CannotConnectThroughSocketOnLocalhost < ConnectionNotEstablished
```
As this error is MySQL2 specific it should be defined in that adapter only

rafaelfranca commented on activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb:
```diff
@@ -18,6 +18,12 @@ def test_connection_error
 
+  def test_connection_error_via_socket_on_localhost
+    assert_raises ActiveRecord::CannotConnectThroughSocketOnLocalhost do
+      ActiveRecord::ConnectionAdapters::Mysql2Adapter.new(host: "localhost", prepared_statements: false).connect!
```
Does this test fails when you have mysql running on your local machine outside a container?

ruyrocha replied:
No, they work as intended, @rafaelfranca:

```shell
ruy@devastator ~/G/r/activerecord (feature/new-exception-when-connecting-to-mysql-running-on-localhost)> bundle exec rake test TEST="test/cases/adapters/mysql2/mysql2_adapter_test.rb"
/home/ruy/.asdf/installs/ruby/3.2.3/bin/ruby -w -I"lib:test" /home/ruy/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/rake-13.1.0/lib/rake/rake_test_loader.rb "test/cases/adapters/mysql2/mysql2_adapter_test.rb"
Using mysql2
Run options: --seed 6656

# Running:

..............S.......

Finished in 1.328510s, 16.5599 runs/s, 40.6470 assertions/s.
22 runs, 54 assertions, 0 failures, 0 errors, 1 skips

You have skipped tests. Run with --verbose for details.
/home/ruy/.asdf/installs/ruby/3.2.3/bin/ruby -w -I"lib:test" /home/ruy/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/rake-13.1.0/lib/rake/rake_test_loader.rb "test/cases/adapters/mysql2/mysql2_adapter_test.rb"
Using mysql2
Run options: --seed 56615

# Running:

...................S..

Finished in 1.300546s, 16.9160 runs/s, 41.5210 assertions/s.
22 runs, 54 assertions, 0 failures, 0 errors, 1 skips
```

Here is what's going on behind the scenes:

```shell
ruy@devastator ~> service mariadb status
Redirecting to /bin/systemctl status mariadb.service
● mariadb.service - MariaDB 10.5 database server
     Loaded: loaded (/usr/lib/systemd/system/mariadb.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: active (running) since Mon 2024-01-29 22:08:55 -03; 14min ago
       Docs: man:mariadbd(8)
             https://mariadb.com/kb/en/library/systemd/
    Process: 169568 ExecStartPre=/usr/libexec/mariadb-check-socket (code=exited, status=0/SUCCESS)
    Process: 169593 ExecStartPre=/usr/libexec/mariadb-prepare-db-dir mariadb.service (code=exited, status=0/SUCCESS)
    Process: 169735 ExecStartPost=/usr/libexec/mariadb-check-upgrade (code=exited, status=0/SUCCESS)
   Main PID: 169703 (mariadbd)
     Status: "Taking your SQL requests now..."
      Tasks: 13 (limit: 18259)
     Memory: 872.6M
        CPU: 8.466s
     CGroup: /system.slice/mariadb.service
             └─169703 /usr/libexec/mariadbd --basedir=/usr

Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: you need to be the system 'mysql' user to connect.
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: After connecting you can set the password, if you would need to be
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: able to connect as any of these users with a password and without sudo
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: See the MariaDB Knowledgebase at https://mariadb.com/kb
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: Please report any problems at https://mariadb.org/jira
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: The latest information about MariaDB is available at https://mariadb.org/.
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: Consider joining MariaDB's strong and vibrant community:
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: https://mariadb.org/get-involved/
Jan 29 22:08:55 devastator (mariadbd)[169703]: mariadb.service: Referenced but unset environment variable evaluates to an empty string: MYSQLD_OPTS, _WSREP_NEW_CLUSTER
Jan 29 22:08:55 devastator systemd[1]: Started mariadb.service - MariaDB 10.5 database server.
ruy@devastator ~> strace -o mysql_connection.txt mysql --user rails --host localhost
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 181
Server version: 10.5.23-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> ^DBye
ruy@devastator ~> rg 'mysql.sock' mysql_connection.txt
279:connect(3, {sa_family=AF_UNIX, sun_path="/var/lib/mysql/mysql.sock"}, 110) = 0
ruy@devastator ~> strace -o mysql_connection.txt mysql --user rails --host 127.0.0.1
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 182
Server version: 10.5.23-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> ^DBye
ruy@devastator ~> rg 'mysql.sock' mysql_connection.txt
ruy@devastator ~ [1]> rg '127.0.0.1' mysql_connection.txt
1:execve("/usr/bin/mysql", ["mysql", "--user", "rails", "--host", "127.0.0.1"], 0xffffc6f8c210 /* 50 vars */) = 0
278:connect(3, {sa_family=AF_INET, sin_port=htons(3306), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
ruy@devastator ~>
```

ruyrocha committed:
Add new exception when attempting to connect to MySQL running on...

 localhost.

Add test to MySQL connection error on localhost via socket.

Update CHANGELOG.

Move exception to MySQL2 adapter.

Address Rubocop warnings.

matthewd commented:
It feels rather awkward to me for us to declare a semi-standard upstream behaviour to be "wrong", and hide the real error because we assume we know better. We shouldn't be telling people to change their connection configuration if it's fine... maybe the server's just not running -- _especially_ if you're pointing at localhost, which is likely in development.

rafaelfranca commented:
Yeah, that is my fear as well. It is possible that you are running the server in a container, but it is also possible that you didn't start the server. If this change makes sense, it should be on MySQL not here.

ruyrocha commented:
As stated earlier, this issue has been around pretty much since 2007, and this is an attempt to be more polite and verbose around the error.

The short path is: use Trilogy. :)
